static/js/pages/CourseSearchPage.js

```diff
@@ -112,415 +63,442 @@ export type CourseSearchParams = {
-    }
+  const loadMore = useCallback(
+    () => {
+      if (!loaded) {
+        // this function will be triggered repeatedly by <InfiniteScroll />, filter it to just once at a time
+        return
+      }
 
-    this.updateSearchState({
-      activeFacets
-    })
-  }
+      runSearch(text, activeFacets, true)
+    },
+    [runSearch, loaded, text, activeFacets]
+  )
 
-  onUpdateFacets = (e: Object) => {
-    this.toggleFacet(e.target.name, e.target.value, e.target.checked)
-  }
+  const clearAllFilters = useCallback(
+    () => {
+      setText(null)
+      setActiveFacets(INITIAL_FACET_STATE)
+    },
+    [setText, setActiveFacets]
+  )
 
-  updateText = async (event: Object) => {
-    const text = event ? event.target.value : ""
-    await this.updateSearchState({ text }, true)
-  }
+  // this effect here basically listens to activeFacets for changes and re-runs
+  // the search whenever it changes. we always want the facet changes to take
+  // effect immediately, so we need to either do this or call runSearch from
+  // our facet-related callbacks. this approach lets us avoid having the
+  // facet-related callbacks (toggleFacet, etc) be dependent on then value of
+  // the runSearch function, which leads to too much needless churn in the
+  // facet callbacks and then causesexcessive re-rendering of the facet UI
+  useEffect(
+    () => {
+      runSearch(text, activeFacets)
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [activeFacets]
+  )
 
-  useSuggestion = async (text: string) => {
-    await this.updateSearchState({ text })
-    this.runSearch()
-  }
+  const toggleFacet = useCallback(
+    async (name: string, value: string, isEnabled: boolean) => {
+      const newFacets = R.clone(activeFacets)
 
-  getFavoriteOrListedObject = (result: LearningResourceResult) => {
-    // Get the latest data from state if any to reflect recent changes in favorites/lists
-    const { entities } = this.props
-    const {
-      courses,
-      programs,
-      userLists,
-      videos,
-      podcasts,
-      podcastEpisodes
-    } = entities
-    switch (result.object_type) {
-    case LR_TYPE_COURSE:
-      return courses ? courses[result.id] || null : null
-    case LR_TYPE_PROGRAM:
-      return programs ? programs[result.id] || null : null
-    case LR_TYPE_USERLIST:
-      return userLists ? userLists[result.id] || null : null
-    case LR_TYPE_VIDEO:
-      return videos ? videos[result.id] || null : null
-    case LR_TYPE_PODCAST:
-      return podcasts ? podcasts[result.id] || null : null
-    case LR_TYPE_PODCAST_EPISODE:
-      return podcastEpisodes ? podcastEpisodes[result.id] || null : null
-    case LR_TYPE_LEARNINGPATH:
-      return userLists ? userLists[result.id] || null : null
-    }
-  }
+      if (isEnabled) {
+        newFacets[name] = _.union(newFacets[name] || [], [value])
+      } else {
+        newFacets[name] = _.without(newFacets[name] || [], value)
```

Comment:
Is it convenient to write a test covering this line? It's missing from code coverage

Reply:
yeah, I can add a test tehre